lib/Dialect/SV/Transforms/HWExportModuleHierarchy.cpp

```diff
@@ -0,0 +1,135 @@
+                           llvm::raw_ostream &os) {
+  os << "{"
+     << "\"instance_name\":\"" << inst.instanceName() << "\","
+     << "\"module_name\":\"" << inst.moduleName() << "\","
+     << "\"instances\":[";
+
+  auto moduleOp =
+      symbolTable.lookup<hw::HWModuleOp>(inst.moduleNameAttr().getValue());
+
+  // Only recurse on module ops, not extern or generated ops, whose internal are
+  // opaque.
+  if (moduleOp) {
+    bool firstIteration = true;
+    for (auto op : moduleOp.getOps<hw::InstanceOp>()) {
+      if (firstIteration)
+        firstIteration = false;
+      else
+        os << ",";
+      printHierarchy(op, symbolTable, os);
+    }
+  }
+
+  os << "]}";
+}
+
+/// Return the JSON-serialized module hierarchy for the given module as the top.
+static std::string extractHierarchyFromTop(hw::HWModuleOp op,
+                                           SymbolTable &symbolTable) {
+  std::string resultBuffer;
+  llvm::raw_string_ostream os(resultBuffer);
+
+  // As a special case for top-level module, set instance name to module name,
+  // since the top-level module is not instantiated.
+  os << "{"
+     << "\"instance_name\":\"" << op.getName() << "\","
+     << "\"module_name\":\"" << op.getName() << "\","
+     << "\"instances\":[";
+
+  bool firstIteration = true;
+  for (auto op : op.getOps<hw::InstanceOp>()) {
+    if (firstIteration)
+      firstIteration = false;
+    else
+      os << ",";
+    printHierarchy(op, symbolTable, os);
+  }
+  os << "]}";
+
+  return resultBuffer;
+}
+
+/// Find the modules corresponding to the firrtl mainModule and DesignUnderTest,
+/// and if they exist, emit a verbatim op with the module hierarchy for each.
+void HWExportModuleHierarchyPass::runOnOperation() {
+  mlir::ModuleOp mlirModule = getOperation();
+  auto builder = OpBuilder::atBlockEnd(mlirModule.getBody());
+  SymbolTable symbolTable(mlirModule);
+
+  if (auto mainModuleName =
+          mlirModule->getAttrOfType<mlir::StringAttr>("firrtl.mainModule")) {
+    auto mainModule = symbolTable.lookup<hw::HWModuleOp>(mainModuleName);
+    assert(mainModule &&
+           "firrtl.mainModule should match the name of some HWModuleOp");
+
+    auto verbatimOp = builder.create<sv::VerbatimOp>(
+        builder.getUnknownLoc(),
+        extractHierarchyFromTop(mainModule, symbolTable));
+    verbatimOp->setAttr(
+        "output_file", hw::OutputFileAttr::get(
+                           builder.getStringAttr(""),
+                           builder.getStringAttr("testharness_hier.json"),
+                           /*exclude_from_filelist=*/builder.getBoolAttr(true),
+                           /*exclude_replicated_ops=*/builder.getBoolAttr(true),
+                           &getContext()));
+  }
+
+  for (auto op : mlirModule.getOps<hw::HWModuleOp>()) {
+    if (op->hasAttr("firrtl.DesignUnderTest")) {
+      auto verbatimOp = builder.create<sv::VerbatimOp>(
+          builder.getUnknownLoc(), extractHierarchyFromTop(op, symbolTable));
```

Comment:
That sounds like a reasonable thing to do in our case here, but I'm not sure if this will translate cleanly to the approach of having a generic "dump hierarchy from here" attribute, since we'd need to know ahead of time which hierarchies completely contain other hierarchies.

Reply:
BTW, I thought about this a bit more, and I realized that we can't actually do this, at least not without introducing some extra handling. The reason is that the JSON strings are not actually 100% composable, and the issue lies in the top-level module. The module hierarchy JSON is really an _instance_ hierarchy, and we have to synthesize a fake top-level instance, where we've chosen to just use the module name as the instance name. However, in the case where we have an outer module (e.g. the test harness) that contains an inner module (e.g. the DUT), we can't just take the inner module's module hierarchy string because the inner module's instance name won't match its actual instance name within the outer module.